core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala

```diff
@@ -0,0 +1,1096 @@
+      context.parent ! staleQueueRemovedMsg
+
+      goto(Removing) using getRemovingData(data, outdated = true)
+
+    case Event(t: FailureMessage, _) =>
+      logging.error(this, s"[$invocationNamespace:$action:$stateName] got an unexpected failure message: $t")
+
+      stay
+
+    case Event(msg: DecisionResults, _) =>
+      val DecisionResults(result, num) = msg
+      result match {
+        case AddInitialContainer if num > 0 =>
+          initialized = true
+          val msgs = generateContainerCreationMessages(num)
+          containerManager ! ContainerCreation(msgs, memory, invocationNamespace)
+
+        case AddContainer if num > 0 =>
+          val msgs = generateContainerCreationMessages(num)
+          containerManager ! ContainerCreation(msgs, memory, invocationNamespace)
+
+        case enable: EnableNamespaceThrottling =>
+          if (num > 0) {
+            val msgs = generateContainerCreationMessages(num)
+            containerManager ! ContainerCreation(msgs, memory, invocationNamespace)
+          }
+          self ! enable
+
+        case DisableNamespaceThrottling =>
+          if (num > 0) {
+            val msgs = generateContainerCreationMessages(num)
+            containerManager ! ContainerCreation(msgs, memory, invocationNamespace)
+          }
+          self ! DisableNamespaceThrottling
+
+        case Pausing =>
+          logging.warn(
+            this,
+            s"[$invocationNamespace:$action:$stateName] The limit value is less than 0. No activation can be handled so the queue becomes the Flushing state.")
+          self ! FailedCreationJob(
+            CreationId.void,
+            invocationNamespace,
+            action,
+            revision,
+            ZeroNamespaceLimit,
+            namespaceLimitUnderZero)
+      }
+      stay
+
+    // this should not happen
+    case otherMsg =>
+      logging.warn(this, s"[$invocationNamespace:$action:$stateName] received unexpected message: $otherMsg")
+
+      stay
+  }
+
+  onTransition {
+    case Uninitialized -> _ => unstashAll()
+    case _ -> Flushing      => setTimer("StopQueue", StateTimeout, queueConfig.flushGrace, repeat = true)
+    case Flushing -> _      => cancelTimer("StopQueue")
+  }
+
+  onTermination {
+    case _ =>
+      // logscheduler must be canceled when FSM is terminated
+      logScheduler.cancel()
+
+      // the lifecycle of DecisionMaker conforms to the one of MemoryQueue
+      actorSystem.stop(decisionMaker)
+  }
+
+  initialize()
+
+  private def cleanUpDataAndGotoRemoved() = {
+    cleanUpWatcher()
+    cleanUpData()
+    context.parent ! queueRemovedMsg
+
+    goto(Removed) using NoData()
+  }
+
+  private def cleanUpActorsAndGotoRemoved(data: FlushingData) = {
+    cleanUpActors(data)
+    cleanUpData()
+
+    context.parent ! queueRemovedMsg
+
+    goto(Removed) using NoData()
+  }
+
+  private def cleanUpActorsAndGotoRemovedIfPossible(data: RemovingData) = {
+    if (queue.isEmpty && requestBuffer.isEmpty) {
```

Comment:
should add `requestBuffer = requestBuffer.filter(!_.promise.isCompleted)` before this line: `if (queue.isEmpty && requestBuffer.isEmpty) {`

This is avoid `when updates queue,  in spite of all requests in requestBuffer will be finished finally, but the completed request may not be removed from requestBuffer`,so need to add `requestBuffer.filter(!_.promise.isCompleted)`